controller-server/src/main/java/com/yahoo/vespa/hosted/controller/deployment/DeploymentTrigger.java

```diff
@@ -275,53 +276,51 @@ private ApplicationController applications() {
     private List<Job> computeReadyJobs(ApplicationId id) {
         List<Job> jobs = new ArrayList<>();
         applications().get(id).ifPresent(application -> {
-            List<Step> productionSteps = application.deploymentSpec().steps().stream()
-                                                    .filter(step -> step.deploysTo(prod) || step.zones().isEmpty())
-                                                    .collect(toList());
-
             Change change = application.changeAt(clock.instant());
             Optional<Instant> completedAt = max(application.deploymentJobs().statusOf(systemTest)
                                                         .<Instant>flatMap(job -> job.lastSuccess().map(JobRun::at)),
                                                 application.deploymentJobs().statusOf(stagingTest)
                                                         .<Instant>flatMap(job -> job.lastSuccess().map(JobRun::at)));
             String reason = "New change available";
-            List<Job> testJobs = Collections.emptyList();
+            List<Job> testJobs = null; // null means "uninitialised", while empty means "don't run any jobs".
 
             if (change.isPresent())
-                for (Step step : productionSteps) {
+                for (Step step : productionStepsOf(application)) {
                     Set<JobType> stepJobs = step.zones().stream().map(order::toJob).collect(toSet());
-                    Map<Instant, List<JobType>> jobsByCompletion = stepJobs.stream().collect(groupingBy(job -> completedAt(change, application, job).orElse(Instant.EPOCH)));
-                    if (jobsByCompletion.containsKey(Instant.EPOCH)) { // Step incomplete because some jobs remain, trigger those if the previous step was done, or required test steps.
-                        for (JobType job : jobsByCompletion.get(Instant.EPOCH)) {
+                    Map<Optional<Instant>, List<JobType>> jobsByCompletion = stepJobs.stream().collect(groupingBy(job -> completedAt(change, application, job)));
+                    if (jobsByCompletion.containsKey(Optional.empty())) { // Step is incomplete; trigger remaining jobs if ready, or their test jobs if untested.
```

Comment:
There was a reason for this change, it confused both me and others reading the code. It even confuses IntelliJs static analysis.

Reply:
Generics type arguments are not inferable, no, because of `Map.get(Object key)` and `Map.containsKey(Object key)`. However, all `Optional.empty()` are equal -- they are _empty_. They hold no value for you to use, and there's no possibility of harm even though IntelliJ complains. 

As for readability, do you really prefer:

1. Sort jobs by time of completion; if they haven't completed, pretend they completed in the 1970s (or a very long time from now)
2. If any jobs completed in the 1970s (or a very long time from now), they didn't really complete, so then ...

to 

1. Sort jobs by optional time of completion.
2. If any jobs have no time of completion, then ...

I'm absolutely certain I prefer the latter, even though it puts an extra type argument in the code. `Instant` has no value for `never`, and using `the 70s` or `the far distant future` as `never` is not a good thing. `Optional<Instant>`, however, has a perfect value for `never`, namely, `empty`. 
The keys aren't even forced into this `Optional` -- they are already `Optional`s from where they are computed. I doubt you dispute the return value of that method.

Reply:
I think the problem here is that the code _is_ hard to read, but it's _not_ because of the type of the keys in this map. If you take a step back, I think you'll agree to that. Mixing in magic-value `Instant`s  doesn't really help -- I think it makes it even worse.

Reply:
Please consider the approach taken in the last commit: `Increase readability at the cost of code duplication`

Reply:
This is a large code base and code should readable by other developers, besides the author. This sometimes requires us to write less esoteric code. I do understand the finer details of why IntelliJ complains, type erasure and your semantics argument, but that's besides the point.

Yesterday proved that this was problematic for others to read, I think that's sufficient reason to improve the code.

Reply:
Great!

Reply:
Of course I agree the code should be readable; I just don't think that was the right change to fix it.